consensus/safety-rules/src/remote.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{ConsensusState, Error, SafetyRules, TSafetyRules};
+use consensus_types::{
+    block::Block, block_data::BlockData, common::Payload, quorum_cert::QuorumCert,
+    timeout::Timeout, vote::Vote, vote_proposal::VoteProposal,
+};
+use libra_types::crypto_proxies::Signature;
+use serde::{Deserialize, Serialize};
+use std::sync::{Arc, RwLock};
+
+#[derive(Deserialize, Serialize)]
+pub enum SafetyRulesInput<T> {
+    ConsensusState,
+    Update(Box<QuorumCert>),
+    StartNewEpoch(Box<QuorumCert>),
+    #[serde(bound = "T: Payload")]
+    ConstructAndSignVote(Box<VoteProposal<T>>),
+    #[serde(bound = "T: Payload")]
+    SignProposal(Box<BlockData<T>>),
+    SignTimeout(Box<Timeout>),
+}
+
+pub struct RemoteService<T> {
+    internal: SafetyRules<T>,
+}
+
+impl<T: Payload> RemoteService<T> {
+    pub fn new(internal: SafetyRules<T>) -> Self {
+        Self { internal }
+    }
+
+    pub fn handle_message(&mut self, input_message: Vec<u8>) -> Result<Vec<u8>, Error> {
+        let input = Error::result(lcs::from_bytes(&input_message))?;
+
+        let output = match input {
+            SafetyRulesInput::ConsensusState => lcs::to_bytes(&self.internal.consensus_state()),
+            SafetyRulesInput::Update(qc) => lcs::to_bytes(&self.internal.update(&qc)),
+            SafetyRulesInput::StartNewEpoch(qc) => {
+                lcs::to_bytes(&self.internal.start_new_epoch(&qc))
+            }
+            SafetyRulesInput::ConstructAndSignVote(vote_proposal) => {
+                lcs::to_bytes(&self.internal.construct_and_sign_vote(&vote_proposal))
+            }
+            SafetyRulesInput::SignProposal(block_data) => {
+                lcs::to_bytes(&self.internal.sign_proposal(*block_data))
+            }
+            SafetyRulesInput::SignTimeout(timeout) => {
+                lcs::to_bytes(&self.internal.sign_timeout(&timeout))
+            }
+        };
+
+        Error::result(output)
+    }
+}
+
+pub struct RemoteClient<T> {
+    service: Arc<RwLock<RemoteService<T>>>,
```

Comment:
lol I assume it's placeholder for now?

Reply:
Hahaha, yes, this is just so I can verify end-to-end. Although I guess one could argue we could have 3 different parameters here:
- Internal
- Threaded
- Process

Each is slightly different because of the expected set of parameters...